Adding controller tests.

vpassapera opened the pull request:
Controller tests.

I tested this with my data sets and it worked but it requires changes to the handlers and ems (on another PR).

The data seeds should follow an interface, below is what I have for now. If there is any issues let me know.

vpassapera commented:
Let me know before we merge this so I can squash it into a single commit. I'm waiting for the hooks to resolve, otherwise travis gets weird about the commit not existing.

vpassapera commented:
CS fixed.

jrjohnson commented on src/Ilios/CoreBundle/Tests/Controller/AamcMethodControllerTest.php:
```diff
@@ -0,0 +1,167 @@
+    {
+        $aamcMethod = $this->container
+            ->get('ilioscore.dataloader.aamcmethod')
+            ->getOne()['aamcMethod']
```
I think ::getOne() should return an array of the data - not an array keyed by 'aamcMethod' which we have to extract.

jrjohnson commented on src/Ilios/CoreBundle/Tests/Controller/AamcMethodControllerTest.php:
```diff
@@ -0,0 +1,167 @@
+            $this->getUrl('post_aamcmethods'),
+            json_encode(
+                $this->container->get('ilioscore.dataloader.aamcmethod')
+                    ->create()['aamcMethod']
```
Same as getOne() I think the create() factory should return an array or data not an array keyed by 'aamcMethod'

jrjohnson commented on src/Ilios/CoreBundle/Tests/Controller/AamcMethodControllerTest.php:
```diff
@@ -0,0 +1,167 @@
+    public function testPostBadAamcMethod()
+    {
+        $invalidAamcMethod = array_shift(
+            $this->container->get('ilioscore.dataloader.aamcmethod')->invalid()
```
Same thing again, lets just return the array we need instead of one with an extra layer.

jrjohnson commented on src/Ilios/CoreBundle/Tests/Controller/AamcMethodControllerTest.php:
```diff
@@ -0,0 +1,167 @@
+        $this->assertJsonResponse($response, Codes::HTTP_CREATED);
+        $this->assertEquals(
+            $this->container->get('ilioscore.dataloader.aamcmethod')
+                ->getLastCreated()['aamcMethod'],
```
I don't like getLastCreated being part of the dataLoader.  I would much rather see us make a second GET request for the updated record and check in that way.  getLastCreated is a DB interaction and the factories should just provide data.

jrjohnson commented on src/Ilios/CoreBundle/Tests/Controller/AamcMethodControllerTest.php:
```diff
@@ -0,0 +1,167 @@
+    {
+        $aamcMethod = $this->container
+            ->get('ilioscore.dataloader.aamcmethod')
+            ->createWithId()['aamcMethod']
```
I don't like createWithId() this seems like a database concern not part of the dataLoaders.  Why not just use getOne() here?

vpassapera replied:
Because for a post, you don't want to create one with an ID, but with a put you do... I am definitely open to suggestions though!

jrjohnson commented on src/Ilios/CoreBundle/Tests/Controller/AamcMethodControllerTest.php:
```diff
@@ -0,0 +1,167 @@
+            ->createWithId()['aamcMethod']
+        ;
+
+        $this->createJsonRequest(
```
Why are we dong a put then delete in the delete test?

vpassapera replied:
To ensure that the entity we are trying to delete actually exists. Persistence doesn't go across each test action. Suggestions welcome!

jrjohnson commented on src/Ilios/CoreBundle/Tests/Controller/AamcMethodControllerTest.php:
```diff
@@ -0,0 +1,167 @@
+        $this->assertEquals($response->getStatusCode(), Codes::HTTP_BAD_REQUEST);
+    }
+
+    public function testPutAamcMethod()
```
PUT is by far the most complicated request because we need to check that it is properly handling relationship updates.  This likely isn't doable with boilerplate and generators so I'm thinking the way to go might be to mark these with 
```php
$this->markTestIncomplete(
  'PUT is not testing relationship updates.'
);
```

For bonus points use the generator to add a commented out list of the relationships for the entity to make it easier to write the custom code later.

vpassapera replied:
I thought that relationships through the owning side would be persisted by doctrine. So doesn't it depend on the data seed?

jrjohnson commented:
Looks great.  My comments above are mostly about the interface for the data loaders.   I don't see an interface defined here - that along with the yml to inject them should be part of this PR.  I think I may be able to auto generate the loaders from our demo data once the interface is complete.  Fixtures probably need to be hand written.

A couple of changes to the interface and an update to the PUT test and this is looking good.  If we still need to team write the data loaders and fixtures then you should put this code into a shared branch on ilios/ilios so we can both commit to it.

vpassapera commented:
I think I fixed / addressed most of the issues. I went ahead and opened a feature/controller-tests branch on the main repo so I'm closing this PR for now and re-opening one from that branch.
